src/json_render.cpp

```diff
@@ -0,0 +1,416 @@
+    vfprintf(ar->f, format, ap);
+    va_end(ap);
+    fprintf(ar->f, "\":");
+}
+void print_json_object_end(JsonRender *ar) {
+    ar->needs_comma.pop();
+
+    if (pretty_json)
+        fprintf(ar->f, " ");
+    fprintf(ar->f, "}");
+}
+void print_json_bool(JsonRender *ar, bool value) {
+    json_begin(ar);
+
+    fprintf(ar->f, value ? "true" : "false");
+}
+void print_json_string(JsonRender *ar, const char *format, ...) {
+    json_begin(ar);
+
+    fprintf(ar->f, "\"");
+    va_list ap;
+    va_start(ap, format);
+    vfprintf(ar->f, format, ap);
+    va_end(ap);
+    fprintf(ar->f, "\"");
+}
+void print_json_string_start(JsonRender *ar) {
+    json_begin(ar);
+
+    fprintf(ar->f, "\"");
+}
+void print_json_string_end(JsonRender *ar) {
+    fprintf(ar->f, "\"");
+}
+void print_json_value(JsonRender *ar, const char *format, ...) {
+    json_begin(ar);
+
+    va_list ap;
+    va_start(ap, format);
+    vfprintf(ar->f, format, ap);
+    va_end(ap);
+}
+
+void print_json_location(JsonRender *ar, AstNode *node) {
+    // TODO: document if we use 0 or 1 based lines.
+    print_json_string(ar, "%s:%d:%d", buf_ptr(node->owner->path), node->line + 1, node->column);
+}
+
+void json_print(CodeGen* codegen, FILE *f, AstNode *node) {
+    JsonRender ar0 = {0};
+    ar0.codegen = codegen;
+    ar0.f = f;
+    JsonRender *ar = &ar0;
+    // for (size_t i = 0; i < codegen->global_vars.length; ++i) {
+    //     TldVar* entry = codegen->global_vars.at(i);
+    //     printf("%s\n", buf_ptr(&entry->var->name));
+    // }
+
+    // for (size_t i = 0; i < codegen->all_types.length; ++i) {
+        // TypeTableEntry *entry = codegen->all_types.at(i);
+        // printf("%s\n", buf_ptr(&entry->name));
+    // }
+    print_json_array_start(ar);
+
+    { auto it = codegen->import_table.entry_iterator(); for (;;) {
+        auto *entry = it.next();
+        if (!entry)
+            break;
+
+        // FIXME: better limiting options
+        if (strcmp(buf_ptr(entry->value->path), "/Users/jdufault/personal/termbox-zig/hello.zig") != 0)
+            continue;
+
+        // printf(" entry path=%s\n", buf_ptr(entry->value->path));
+        // if (entry->value->decls_scope->container_type)
+            // printf("%s\n", buf_ptr(&entry->value->decls_scope->container_type->name));
+
+        { auto decl_table_it = entry->value->decls_scope->decl_table.entry_iterator(); for (;;) {
+            auto* decl_table_entry = decl_table_it.next();
+            if (!decl_table_entry)
+                break;
+
+            print_json_object_start(ar);
+
+            print_json_object_key(ar, "name");
+            print_json_string(ar, "%s", buf_ptr(decl_table_entry->value->name));
+
+            print_json_object_key(ar, "id");
+            print_json_string(ar, tld_id_str(decl_table_entry->value->id));
+
+            switch (decl_table_entry->value->id) {
+                case TldIdVar:
+                    {
+                        auto *var = (TldVar *)decl_table_entry->value;
```

Comment:
It seems like var is not fully populated here, ie, var->var is always null. Any idea why?

Reply:
Seems like this might be because the variable is not resolved (ie, the TldResolution value is unresolved). It looks like resolution happens in codegen_build, which the JSON driver in main.cpp is calling - any idea why the code is not being resolved? Here is how I'm testing:

```
const Aaaaaaaa = struct {
  x: u8
};

pub fn main() void {
  var a = Aaaaaaaa { .x = 3 };
  const std = @import("std");
  std.debug.warn("{}\n", a.x);
}
```

Reply:
I pushed changes to your branch merging conflicts with master. I tried testing the code with this file:

```zig
pub fn main() void {
}
```

and the command: `zig json test.zig`

the result was an empty array: `[  ]`.

what have you been using to test?

Reply:
Sorry, I didn't take into account your comment above. I tried with

```zig
const Aaaaaaaa = struct {
  x: u8
};

pub fn main() void {
  var a = Aaaaaaaa { .x = 3 };
  const std = @import("std");
  std.debug.warn("{}\n", a.x);
}
```

but `zig json` still gives an empty output. is that what you've been getting so far?

Reply:
I pushed a version that I rebased onto master and tested. Can you try again? Maybe there was an issue with the merge.

I'm running like this

```zig json hello.zig```

where hello.zig is the file above.

Reply:
I found the issue (I hard-coded the path to restrict output) - it should work now, though everything is still unresolved.

Reply:
Try adding this to your test file:

```zig
export fn foo() void {
    var a: Aaaaaaaa = undefined;
}
```

you can see that foo as well as Aaaaaaa are now resolved.

looks like the code in build-exe that detects we should add std/special/bootstrap.zig to the compilation isn't triggering when you do `zig json`.

```c++
    if (!g->is_test_build && g->zig_target.os != OsFreestanding &&
        !g->have_c_main && !g->have_winmain && !g->have_winmain_crt_startup &&
        ((g->have_pub_main && g->out_type == OutTypeObj) || g->out_type == OutTypeExe))
    {
        g->bootstrap_import = add_special_code(g, create_bootstrap_pkg(g, g->root_package), "bootstrap.zig");
    }
```

I bet it's the `out_type`

```c++
            } else if (strcmp(arg, "json") == 0) {
                cmd = CmdJson;
                out_type = OutTypeJson;
```

I think that instead of trying to make `zig json` act like `zig build-exe`, we should modify the lazy analysis code to check every top level declaration when `out_type == OutTypeJson`. We will do something similar to this for autogenerated documentation.

Reply:
Where you would add code to check all top level decls:

analyze.cpp:3066 in 
`static void add_top_level_decl(CodeGen *g, ScopeDecls *decls_scope, Tld *tld)`

```c++
    if (is_export) {
        g->resolve_queue.append(tld);
```
 move that to outside the `if`.